### PR TITLE
Feat error improve + proof API improvement

### PR DESCRIPTION
**Pull request by thomas-waite**
## Summary
This PR implements 3 discrete blocks of functionality. Firstly, it performs error handling improvements for the JavaScript (JS) proof construction and verification code (aztec.js and dev-utils packages). Secondly, it implements proof API standardisation (aztec.js) and thirdly, removes WET code i.e. reduces code duplication between proofs (aztec.js).

## Motivation
Error handling is a core functionality of the aztec.js package, with it's primary aim being to prevent the construction and verification of invalid proofs. This is what the current implementation does.

However, it is also desirable for the error handling to perform additional functionality, depending on whether it is proof construction or verification that is being performed.

**Proof construction**
Proof construction is performed by the user, it is client side. It is important that an error is thrown and the transaction aborted immediately, to stop the user submitting (and paying gas costs!) to the smart contract verifier for a faulty proof. Hence, the current model is maintained and extended with more detailed errors and additional checks.

**Proof verification**
This is used primarily for testing the proof verification algo - in the real world deployment it is a smart contract that actually provides verification. To aid more thorough testing of the verification algo, it is desirable to collect **all** potential errors that are thrown and only at the end of the verification step, throw the transaction if it is faulty. It is also desirable, for testing purposes, to be able to use stubs for certain checks in the verification code. This is what this PR does.

Proof API standardisation across all proofs is performed to make the developer UI more intuitive and predictable. It also reduces substantial code duplication, moving the various helper functions to a single proofUtils module. This should make the module easier to work with and build on.

## Implementation
The most major changes apply to the aztec.js and dev-utils packages. dev-utils now contains an errors.js script and a series of standardised ERROR_TYPES in constants.js . 

The aztec.js package now makes use of these in the proof folder, specifically in the index.js scripts of the various proofs (this contains proof construction code) and the verifier.js scripts of the various proofs (proof verification). 

All helpers.js scripts for the proofs have been removed, and replaced with a single proofUtils.js in the proof folder. 

The proof verification code has slightly changed. An errors array is now defined at the start of the proof verification and any errors that are encountered during execution are pushed into this array.At the end of proof verification, this is checked and if errors exist the transaction is aborted.  

Various imports and testing patterns were updated throughout to be compatible with the new error and API patterns. 

### aztec.js
Has most of the major changes. High level changes:
1) Removed all helpers.js scripts from proof/bilateralSwap, proof/dividendComputation, 
2) Added in a proofUtils.js script that contains all the helpers.js functionality

Specific file changes:
**proof/bilateralSwap/index.js, proof/dividendComputation/index.js** 
Now uses the customError() method and ERROR_TYPES to throw more informative error messages

**proof/bilteralSwap/verifier.js, proof/dividendComputation/verifier.js**
Now uses an error array to record all errors thrown

### dev-utils
Now contains an errors.js script and has a set of ERROR_TYPES defined in constants.js

### protocol
Updated testing patterns to be compatible with updated error handling patterns

## Checklist

- [x] Prefix PR title with [WIP] if necessary.
- [ -] Write tests for added code when needed.
- [ -] Update documentation when needed.

### CODE REVIEW

File: packages/aztec.js/test/proof/bilateralSwap/verifier.spec.js [hunk 1]
```diff
@@ -67,73 +65,75 @@ describe('AZTEC bilateral swap verifier tests', () => {
...
         let sender;
+        let parseInputs;
 
         beforeEach(() => {
-            testNotes = helpers.makeTestNotes([10, 20], [10, 20]);
+            // to test failure states we need to pass in bad data to verifier
+            // so we need to turn off proof.parseInputs
+            parseInputs = sinon.stub(proofUtils, 'parseInputs').callsFake(() => { });
+
+            testNotes = proofUtils.makeTestNotes([10, 20], [10, 20]);
```
Comment (ArnSch) [inline packages/aztec.js/test/proof/bilateralSwap/verifier.spec.js]: I think it makes sense to move this to the first test to align with structure of the other tests.

File: packages/aztec.js/test/proof/bilateralSwap/proof.spec.js [hunk 2]
```diff
@@ -1,32 +1,21 @@
...
-}
-
-
 describe('AZTEC bilateral swap proof construction tests', () => {
     let testNotes;
     let sender;
 
     beforeEach(() => {
-        testNotes = helpers.makeTestNotes([10, 20], [10, 20]);
+        testNotes = proofUtils.makeTestNotes([10, 20], [10, 20]);
```
Comment (ArnSch) [inline packages/aztec.js/test/proof/bilateralSwap/proof.spec.js]: Same comment as on previous code reviews, better to have all state for tests be contained in the test case, even if slightly wordier.

File: packages/dev-utils/src/errors.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+const catchAsync = (fn) => {
```
Comment (ArnSch) [inline packages/dev-utils/src/errors.js]: Do we use this anywhere?

Comment (thomas-waite) [reply packages/dev-utils/src/errors.js]: No we don't, good catch

File: packages/aztec.js/test/proof/bilateralSwap/proof.spec.js (hunk 2 above)
Comment (thomas-waite) [reply packages/aztec.js/test/proof/bilateralSwap/proof.spec.js]: Good point, this is a legacy pattern - will update/remove this pattern as I come across it in the codebase

File: packages/aztec.js/test/proof/bilateralSwap/verifier.spec.js (hunk 1 above)
Comment (thomas-waite) [reply packages/aztec.js/test/proof/bilateralSwap/verifier.spec.js]: Done

File: packages/aztec.js/src/proof/proofUtils.js [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+/**
+ * Checks the number of notes, records error if wrong
+ * but does not throw
+ * @method checkNumNotesNoThrow
+ * @memberof module:proofUtils
+ * @param {Object[]} notes - array of AZTEC notes
+ * @param {integer} numNotes - desired number of notes
+ * @param {string[]} errors - record of all errors that are thrown
+ */
+proofUtils.checkNumNotesNoThrow = (notes, numNotes, errors) => {
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/proofUtils.js]: Good idea, done

File: packages/aztec.js/src/proof/bilateralSwap/index.js [hunk 5]
```diff
@@ -1,74 +1,59 @@
...
+    notes.forEach((note) => {
+        const gammaOnCurve = bn128.curve.validate(note.gamma); // checking gamma point
+        const sigmaOnCurve = bn128.curve.validate(note.sigma); // checking sigma point
+
+        if ((gammaOnCurve === false) || (sigmaOnCurve === false)) {
+            throw customError(
+                ERROR_TYPES.NOT_ON_CURVE,
+                {
+                    data: `Is gamma on the curve?: ${gammaOnCurve}
+                    Is sigma on the curve?: ${sigmaOnCurve}
```
[the remark this answers is not in the document]
Comment (thomas-waite) [reply packages/aztec.js/src/proof/bilateralSwap/index.js]: Done

**Pull request merged**